### ISSUE TITLE
cmd/compile: cannot index pointer to generic array

### ISSUE THREAD

**Issue by josharian** [golang/go#49447]
This does not compile:

```go
package p

type a1 [16]byte
type a2 [32]byte

type A interface {
	a1 | a2
}

func elem0[T A](t *T) *byte {
	return &t[0]
}
```

Error: `./x.go:11:10: invalid operation: cannot index t (variable of type *T)`

This shows up in practice if you want to embed an array in a generic struct to avoid extra indirection/allocs, like this:

```go
type S struct {
  // other fields
  arr A
}
```

but pass pointers to the array around internally to avoid making lots of copies of it (particularly as the array gets large).

The equivalent non-generic code compiles:

```go
func elem0concrete(t *a1) *byte {
	return &t[0]
}
```

**Comment by ianlancetaylor** [comment]
CC @griesemer @findleyr 

I think this may be expected at present.

**Comment by griesemer** [comment]
This is currently as expected. Note that the existing rules in the spec require a pointer to an array. Expanding this to type parameters requires that we have pointers to arrays in the type parameter type set (which is accepted). We'd need a new rule to generalize this also across a type parameter boundary. I'm not saying we can't do that, but it's an exception to the otherwise relatively regular rules.

This works:
```Go
package p

type a1 [16]byte
type a2 [32]byte

type A interface {
	*a1 | *a2
}

func elem0[T A](t T) *byte {
	return &t[0]
}
```
is that not sufficient?

**Comment by josharian** [comment]
> is that not sufficient?

This is what I mentioned above about embedding an array in a generic struct to avoid extra indirections and allocs. `struct { x [16]byte }` is importantly different than `struct { x *[16]byte }`.

**Comment by ianlancetaylor** [comment]
Can you show us more context?  Show us the program that doesn't work with @griesemer 's suggestion.

**Comment by josharian** [comment]
In the course of writing a minimized program, I found a workaround:

```
func elem0[T A](t *T) *byte {
	switch a := any(t).(type) {
	case *a1:
		return &a[0]
	case *a2:
		return &a[1]
	default:
		panic("unreachable")
	}
}
```

It's ugly, but it works.

**Comment by findleyr** [comment]
Just catching up and curious about this.

@josharian why is that type switch necessary? Why can't you use @griesemer's suggestion of including the pointer-to-array types in the type parameter type set?

```
type A interface {
	*a1 | *a2
}

func elem0[T A](t T) *byte {
	return &t[0]
}
```

But also, @griesemer I'm not sure that supporting this would make the spec less regular, and supporting this seems straightforward (I have a CL that does it...). I actually think that @josharian demonstrated that it would be less surprising if this worked, than if it doesn't.

**Comment by griesemer** [comment]
I agree that supporting it is straight-forward. But we need an extra rule. We may also need to look into other situations where we may need to apply this (pointer-to-struct in some cases).

**Comment by josharian** [comment]
The use case is a trie for looking up IP addresses without zones. For space reasons, it'd be nice to have an IPv4 trie `[4]byte` and an IPv6 trie `[16]byte`. Generics seemed like an obvious fit. This is performance sensitive, so it is important that the array be a struct field, rather than a pointer to an array (data locality, fewer allocs, no nil checks).

I was definitely using the mental heuristic you indicated.

I also just encountered an identical issue with slicing:

```go
func asSlice[T A](t T) []byte {
	return t[:]
}
```

also does not compile, even though it compiles for each of the types in T.

Here's a more complete program:

```go
package p

type a1 [16]byte
type a2 [32]byte

type A interface {
	a1 | a2
}

type S[T A] struct {
	arr T // initial backing array for buf
	buf []byte
}

func makeT[T A]() *S[T] {
	var s S[T]
	switch a := any(s.arr).(type) {
	case a1:
		s.buf = a[:]
	case a2:
		s.buf = a[:]
	default:
		panic("unreachable")
	}
	return &s
}
```

It'd be nice to be able to remove the dynamic type switch from `makeT` and just unilaterally write `s.buf = a[:]`.
